File: dotnet/src/Core/Logging/EventLoggerFactory.cs
using Microsoft.Extensions.Logging;

namespace Agience.Core.Logging
{
    public class EventLoggerFactory : ILoggerFactory
    {
        private readonly List<ILoggerProvider> _providers = new();
        private readonly string? _agentId;

        public EventLoggerFactory(string? agentId)
        {
            _agentId = agentId;
        }

        public ILogger CreateLogger(string categoryName)
        {
            return CreateScopedLogger<ILogger>(categoryName, _agentId);
        }

        public ILogger<T> CreateLogger<T>()
        {
            return CreateScopedLogger<ILogger<T>>(typeof(T).FullName ?? typeof(T).Name, _agentId);
        }

        private TLogger CreateScopedLogger<TLogger>(string categoryName, string? agentId) where TLogger : ILogger
        {
            var loggers = _providers.Select<ILoggerProvider, ILogger>(provider =>
                provider is EventLoggerProvider agienceProvider
                    ? agienceProvider.CreateLogger<TLogger>(agentId)
                    : (TLogger)provider.CreateLogger(categoryName)
            ).ToList();

            var scopeData = new Dictionary<string, object>();

            if (!string.IsNullOrEmpty(agentId))
            {
                scopeData["AgentId"] = agentId;
            }

            return (TLogger)Activator.CreateInstance(typeof(ScopedCompositeLogger<TLogger>), loggers, scopeData)!;
        }

        public void AddProvider(ILoggerProvider provider)
        {
            _providers.Add(provider);
        }

        public void Dispose()
        {
            foreach (var provider in _providers)
            {
                provider.Dispose();
            }
            _providers.Clear();
        }
    }
}

File: core/sdk/python/logging/event_logger_factory.py
from typing import TypeVar, Generic, List, Dict, Any, Optional, Type
from abc import ABC, abstractmethod
from .event_logger import ILogger
from .event_logger_provider import EventLoggerProvider
from .scoped_composite_logger import ScopedCompositeLogger

# Generic type variable for logger types
TLogger = TypeVar('TLogger', bound=ILogger)
T = TypeVar('T')


class ILoggerProvider(ABC):
    @abstractmethod
    def create_logger(self, category_name: str) -> ILogger:
        pass

    @abstractmethod
    def dispose(self) -> None:
        pass


class ILoggerFactory(ABC):
    @abstractmethod
    def create_logger(self, category_name: str) -> ILogger:
        pass

    @abstractmethod
    def create_logger_generic(self, T: Type) -> ILogger:
        pass

    @abstractmethod
    def add_provider(self, provider: ILoggerProvider) -> None:
        pass

    @abstractmethod
    def dispose(self) -> None:
        pass


class EventLoggerFactory(ILoggerFactory):
    def __init__(self, agent_id: Optional[str]):
        self._providers: List[ILoggerProvider] = []
        self._agent_id = agent_id

    def create_logger(self, category_name: str) -> ILogger:
        return self._create_scoped_logger(ILogger, category_name, self._agent_id)

    def create_logger_generic(self, T: Type) -> ILogger:
        category_name = T.__module__ + '.' + T.__name__
        return self._create_scoped_logger(ILogger, category_name, self._agent_id)

    def _create_scoped_logger(self, logger_type: Type[TLogger],
                              category_name: str,
                              agent_id: Optional[str]) -> TLogger:
        loggers: List[ILogger] = []

        for provider in self._providers:
            if isinstance(provider, EventLoggerProvider):
                logger = provider.create_logger_generic(logger_type, agent_id)
            else:
                logger = provider.create_logger(category_name)
            loggers.append(logger)

        scope_data: Dict[str, Any] = {}
        if agent_id:
            scope_data["AgentId"] = agent_id

        return ScopedCompositeLogger(loggers, scope_data)  # type: ignore

    def add_provider(self, provider: ILoggerProvider) -> None:
        self._providers.append(provider)

    def dispose(self) -> None:
        for provider in self._providers:
            provider.dispose()
        self._providers.clear()


# Example usage
# def create_logger_factory(agent_id: Optional[str]) -> EventLoggerFactory:
#     factory = EventLoggerFactory(agent_id)
#     factory.add_provider(EventLoggerProvider())
#     return factory


# Example usage
# factory = create_logger_factory("agent1")
# logger = factory.create_logger("CategoryName")
# typed_logger = factory.create_logger_generic(str)  # For ILogger<string> equivalent

# try:
#     # Use the logger
#     pass
# finally:
#     factory.dispose()
